src/wui/info_panel.cc

```diff
@@ -0,0 +1,528 @@
+		} else {
+			display_mode_ = DisplayMode::kOnMouse_Hidden;
+			set_textareas_visibility(false);
+		}
+		break;
+	default:
+		NEVER_HERE();
+	}
+
+	layout();
+}
+
+void InfoPanel::set_toolbar(MainToolbar& t) {
+	assert(!toolbar_);
+	toolbar_ = &t;
+	toolbar_->on_top = on_top_;
+	toolbar_->draw_background = false;
+	layout();
+}
+
+inline bool InfoPanel::is_mouse_over_panel() const {
+	uint8_t h;
+	switch (display_mode_) {
+	case DisplayMode::kMinimized:
+		return false;
+	case DisplayMode::kOnMouse_Hidden:
+		h = kSpacing;
+		break;
+	default:
+		h = MainToolbar::kButtonSize;
+		break;
+	}
+	return last_mouse_pos_.x >= 0 && last_mouse_pos_.x <= get_w() &&
+	       (on_top_ ? (last_mouse_pos_.y >= 0 && last_mouse_pos_.y <= h) :
+	                  (last_mouse_pos_.y <= get_h() && last_mouse_pos_.y >= get_h() - h));
+}
+
+void InfoPanel::set_textareas_visibility(bool v) {
+	for (UI::Textarea* t : {&text_time_speed_, &text_fps_, &text_coords_}) {
+		t->set_visible(v);
+	}
+	if (toolbar_) {
+		toolbar_->draw_background = !v;
+	}
+}
+
+bool InfoPanel::handle_mousemove(uint8_t, int32_t x, int32_t y, int32_t, int32_t) {
+	last_mouse_pos_ = Vector2i(x, y);
+
+	switch (display_mode_) {
+	case DisplayMode::kMinimized:
+		return false;
+	case DisplayMode::kPinned:
+		return is_mouse_over_panel();
+	case DisplayMode::kOnMouse_Visible:
+		if (!is_mouse_over_panel()) {
+			display_mode_ = DisplayMode::kOnMouse_Hidden;
+			set_textareas_visibility(false);
+			layout();
+			return false;
+		}
+		return is_mouse_over_panel();
+	case DisplayMode::kOnMouse_Hidden:
+		if (is_mouse_over_panel()) {
+			display_mode_ = DisplayMode::kOnMouse_Visible;
+			set_textareas_visibility(true);
+			layout();
+			return true;
+		}
+		return false;
+	default:
+		NEVER_HERE();
+	}
+}
+
+// Consume mouse click/move events on the panel
+bool InfoPanel::handle_mousepress(uint8_t, int32_t x, int32_t y) {
+	last_mouse_pos_ = Vector2i(x, y);
+	return is_mouse_over_panel();
+}
+bool InfoPanel::handle_mouserelease(uint8_t, int32_t x, int32_t y) {
+	last_mouse_pos_ = Vector2i(x, y);
+	return is_mouse_over_panel();
+}
+
+size_t InfoPanel::index_of(const MessagePreview* mp) const {
+	size_t i = 0;
+	for (const MessagePreview* m : messages_) {
+		if (m == mp) {
+			return i;
+		}
+		++i;
+	}
+	NEVER_HERE();
+}
+
+void InfoPanel::log_message(const std::string& message) {
+	push_message(new MessagePreview(this, message, ""));
+}
+
+void InfoPanel::pop_message(MessagePreview* m) {
+	m->set_visible(false);
+	m->die();
+	for (auto it = messages_.begin(); it != messages_.end(); ++it) {
+		if (*it == m) {
+			messages_.erase(it);
+			layout();
+			return;
+		}
+	}
+	NEVER_HERE();
+}
+
+void InfoPanel::push_message(MessagePreview* message) {
+	messages_.push_back(message);
+
+	if (messages_.size() > kMaxMessages) {
+		pop_message(messages_.front());
+	}
+
+	layout();
+}
+
+void InfoPanel::set_fps_string(const std::string& long_string, const std::string& short_string) {
+	if (get_w() < 970) {
```

Comment:
This is a "magic number". If the string assembly is directly above, it is clearer, where the number comes from.
```suggestion
void InfoPanel::set_fps_string(bool is_debug, double fps, double fps_avg) {
       std::string short_string = "";
       std::string long_string = "";
       if (is_debug) {
	      long_string = (boost::format("%5.1f fps (avg: %5.1f fps)") % fps % fps_avg()).str();
	      short_string = (boost::format("%.1f / %.1f") % fps % fps_avg()).str();
	}
	if (get_w() < 970) {
```